apm-sniffer/apm-sdk-plugin/dubbo-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/asf/dubbo3/DubboInterceptor.java

```diff
@@ -151,6 +137,19 @@ private void dealException(Throwable throwable) {
         span.log(throwable);
     }
 
+    /**
+     * to judge if current is in provider side.
+     *
+     * @param rpcInvocation
+     * @return
+     */
+    private static boolean isProvider(RpcInvocation rpcInvocation) {
+        Invoker<?> invoker = rpcInvocation.getInvoker();
+        return invoker.getUrl()
+                .getParameter("side", "consumer")
+                .equals("provider");
```

Comment:
Is this an official way? It seems strange. I am waiting for Dubbo team to response.

Reply:
We can wait for the Dubbo team...
I write this from my understanding to the scope of `Rpc*Context` and `RpcInvocation`. 
 The `RpcServiceContext#isConsumerSide()` implements like:
```java
public boolean isConsumerSide() {
        return this.getUrl().getSide("provider").equals("consumer");
}
```

Reply:
If this is already official, could we use `RpcServiceContext#isConsumerSide()` directly?

Reply:
The problem is `RpcServiceContext ` not been reset when it's role switched, but `RpcInvocation` is always correctly bounded to the current request or serve request

Reply:
Oh, I see. Could you add this as comments on the codes? This is important to keep this in mind why we change now.

Reply:
OK. And I know the problem, Dubbo will wrap exception so deal with it here(`#afterMethod`) have no problem.
It log error twice because both `MonitorFilter` and `MonitorClusterFilter` invoked when it is consumer side. It look we enhance classes not precisely enough, here is the stack:
```
afterMethod:120, DubboInterceptor (org.apache.skywalking.apm.plugin.asf.dubbo3)
**intercept:97, InstMethodsInter (org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance)
invoke:-1, MonitorFilter (org.apache.dubbo.monitor.support)**
invoke$accessor$Jzsj124D:-1, MonitorClusterFilter (org.apache.dubbo.monitor.support)
call:-1, MonitorClusterFilter$auxiliary$xlYbM1kf (org.apache.dubbo.monitor.support)
**intercept:86, InstMethodsInter (org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance)
invoke:-1, MonitorClusterFilter (org.apache.dubbo.monitor.support)**
invoke:84, FilterChainBuilder$FilterChainNode (org.apache.dubbo.rpc.cluster.filter)
```

We can filter the `method.getDeclaringClass()` if it's `MonitorClusterFilter` to avoid this, if no more graceful way.

Reply:
Agree, targeting a specific one class seems a good idea to avoid duplicated logging.  But the original `try/catch` for `dealException` method itself seems not reasonable, WDYT?

Reply:
Seems no need to `try/catch`, just deal the `result.getException` look ok.

Reply:
+1 from me.

Reply:
I will simplify it in 3.x plugin and run the test. 
There are more problem beyond these, the peer of Span is not correct, ip is `null` when in consumer side and port is `0` in provider side. I will try to fix it today.

Reply:
For tracing perspective, only consumer side span's peer matters. But it is strange to be `null`.

Reply:
@wu-sheng Can you have a look at this, I want to filter by `objInst`'s class name, but found it's still `MonitorClusterFilter`. The stack is below, from my understanding, we are now intercepting `MonitorFilter` nested in `MonitorClusterFilter`, should the `objInst` be `MonitorFilter` class? How did SW internal works here?
![image](https://user-images.githubusercontent.com/4892550/155493223-3935baac-a1e8-400e-b799-f214b778bd48.png)

Reply:
Check instrumentation definition, `DubboInstrumentationBase`.

```java
    public static final String CONSUMER_ENHANCE_CLASS = "org.apache.dubbo.monitor.support.MonitorClusterFilter";
```

Reply:
I see `MonitorClusterFilter` is extends `MonitorFilter` and `invoke()` method is defined in parent `MonitorFilter` class, at this time SW will enhance `MonitorFilter#invoke()` method twice? One for the subclass, another for parent class, as we specified both `MonitorClusterFilter#invoke()` and `MonitorFilter#invoke()`.

Reply:
`MonitorClusterFilter` is a shell from the source code level. So, you only could intercept `MonitorFilter#invode`, no matter how you define this.

Reply:
I don't know why here[https://github.com/apache/skywalking-java/pull/73#issuecomment-987712730](url) says it fails as Consumer not invoking `MonitorFilter`. As I see, logically, the Consumer did not use `MonitorFilter` and it uses `MonitorClusterFilter`, but from our class bytes perspective, both class bytes are defined in `MonitorFilter#invoke()`, we just need to enhance it once. Am I in the right way?

Reply:
OK, same as I see. I will modify it and pass the test.

Reply:
Yes, we just need to declare the parent class only.

Reply:
But I can't see the link from your comments, but the conclusion is certain.

Reply:
https://github.com/apache/skywalking-java/pull/73#issuecomment-987712730
I reduced 2 `Instrumentation` to 1 and testing it.

Reply:
Got it. Let's see.